### ISSUE TITLE
fusion bug exposed by test_kde_example

### ISSUE THREAD

**Issue by ninegua** [IntelLabs/numba#49]
The bug causes this test to fail by producing wrong output, and it is triggered by the shortcut that allowed `ravel` to return the input array directly. This particular test program produces a nested parfor, and when the outer parfor body is being compiled separately during lowering, it triggers more fusion (recursive fusion should have caught this before the lowering, but it missed because full parfor pipeline like simplifyCFG wasn't run during recursion).

The fusion wrongly merged a reduction (from `min`) with a subsequent parfor computation that requires the result of the `min`, which looks like a bug in dependency checking.

**Comment by ninegua** [comment]
A minimum test case is like this:

```
import numpy as np
from numba import njit

def f(A):
    m = np.sum(A)
    return A + m

a=np.ones(10)
print(njit(parallel=True)(f)(a), f(a))
```

**Comment by ehsantn** [comment]
Thank you, @ninegua . Seems like the dependency check code assumes simple maps only. I'm surprised we haven't seen issues before. Working on more general dependency checks.

**Comment by ehsantn** [comment]
Fixed with 700e6cc0e03e11a41b856c1408d4a95c8737f5d4. There are more corner cases that we need to think about though.
